nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/LayerOpValidation.java

```diff
@@ -1466,5 +1471,183 @@ public void testBiasAdd_nchw_nhwc() {
+                        .directionMode(LSTMDirectionMode.FWD)
+                        .gateAct(LSTMActivations.SIGMOID)
+                        .cellAct(LSTMActivations.TANH)
+                        .outAct(LSTMActivations.TANH)
+                        .retFullSequence(true)
+                        .retLastC(true)
+                        .retLastH(true)
+                        .build()
+        ).getAllOutputs();
+
+        assertEquals(3, outputs.size());
+
+        // [bS, nOut, sL]   when directionMode <= 2 && dataFormat == 2 (NST)
+        long[] out = new long[]{miniBatchSize, numUnits};
+        //  [bS, nOut]   when directionMode <  2
+        long[] hL = new long[]{miniBatchSize, numUnits};
+        long[] cL = new long[]{miniBatchSize, numUnits};
+
+
+        assertArrayEquals(out, outputs.get(0).getShape());
+        assertArrayEquals(hL, outputs.get(1).getShape());
+        assertArrayEquals(cL, outputs.get(2).getShape());
+
+
+
+
+
+    }
+
+
+    @Test
+    public void LSTMLayerTestCase2() {
+
+        int miniBatchSize = 10;
+        int numUnits = 10;
+        int maxTSLength = 64;
+        int sL = 10; //small just for test
+
+
+        SameDiff sd = SameDiff.create();
+        // [sL, bS, nIn]  when dataFormat == TNS
+        SDVariable in =sd.var("in", Nd4j.rand( sL, miniBatchSize, numUnits));
+
+
+        SDVariable cLast = sd.var("cLast", Nd4j.zeros(DataType.FLOAT, miniBatchSize, numUnits));
+        SDVariable yLast = sd.var("yLast", Nd4j.zeros(DataType.FLOAT, miniBatchSize, numUnits));
+
+
+        List<SDVariable> outputs = sd.rnn.lstmLayer(
+                maxTSLength, in, cLast, yLast,
+                LSTMLayerWeights.builder()
+                        .weights(sd.var("weights", Nd4j.rand(numUnits, 4 * numUnits)))
+                        .rWeights(sd.var("rWeights", Nd4j.rand(numUnits, 4 * numUnits)))
+                        .build(),
+                LSTMLayerConfig.builder()
+                        .lstmdataformat(LSTMDataFormat.TNS)
+                        .directionMode(LSTMDirectionMode.FWD)
+                        .gateAct(LSTMActivations.SIGMOID)
+                        .cellAct(LSTMActivations.TANH)
+                        .outAct(LSTMActivations.TANH)
+                        .retFullSequence(true)
+                        .retLastC(false)
+                        .retLastH(false)
+                        .build()
+        ).getAllOutputs();
+
+        assertEquals(1, outputs.size());
+
+
+        // 1) [sL, bS, nOut]    when directionMode <= 2 && dataFormat == 0 TND
+        long[] out = new long[]{sL, miniBatchSize, numUnits};
+        //  [bS, nOut]   when directionMode <  2
+        long[] hL = new long[]{miniBatchSize, numUnits};
+        long[] cL = new long[]{miniBatchSize, numUnits};
+
+
+        assertArrayEquals(out, outputs.get(0).getShape());
+        assertArrayEquals(hL, outputs.get(1).getShape());
+        assertArrayEquals(cL, outputs.get(2).getShape());
```

Comment:
Test needs to be fixed - you (correctly) assert 1 output on line 1578, yet then test/expect 3 outputs.

Same thing in the test case below.